internal/validator/component_registry.go

```diff
@@ -0,0 +1,45 @@
+package validator
+
+import (
+	"strings"
+
+	"github.com/grafana/alloy/internal/component"
+	"github.com/grafana/alloy/syntax/ast"
+)
+
+var _ component.Registry = (*componentRegistry)(nil)
+
+func newComponentRegistry(cr component.Registry) *componentRegistry {
+	return &componentRegistry{
+		parent: cr,
+		custom: make(map[string]component.Registration),
+	}
+}
+
+// componentRegistry wraps a component.Registry and is used to register
+// custom components (declare blocks).
+type componentRegistry struct {
+	parent component.Registry
+	custom map[string]component.Registration
+}
+
+func (cr *componentRegistry) Get(name string) (component.Registration, error) {
+	parts := strings.Split(name, ".")
+	// Custom component registered with their label and should be the first part of the name.
+	// This does hower.
+	// FIXME(kalleep): right now we register modules as a custom component and only validate the
+	// the namespace part. We cant really know what components is contained within a module without
+	// importing in first. Maybe we could have a option for validation to resolve modules so we could
+	// validate them correctly but by default we don't resolve and validate them.
```

Comment:
```suggestion
	// FIXME(kalleep): right now we register modules as custom components and only validate the
	// namespace part. We can't really know what components are contained within a module without
	// importing it first. Maybe we could have an option for validation to resolve modules so we could
	// validate them correctly but by default we don't resolve and validate them.
```

I agree with the idea of the option for modules!